// mext-render/lib/render/index.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

const html_1 = require('../html');

const document_1 = require('./document');

const app_1 = require('./app');

exports.default = async (ctx, renderReactToString) => {
  // render时间log
  const _startTime = Number(new Date());

  ctx.mext.RootComponent = app_1.default(ctx);

  try {
    // 根据cache判断返回html还是数据结构
    return await html_1.default(
      ctx,
      document_1.default({
        url: ctx.request.url,
        mode: ctx.mext.config.mode,
        hashName: ctx.mext.config.hashName,
        router: ctx.mext.config.router,
        dev: ctx.mext.dev,
        map: ctx.mext.map,
        manifest: ctx.mext.manifest,
        cache: ctx.mext.cache,
        initialState: ctx.mext.initialState,
        assetPrefixs: ctx.mext.config.assetPrefixs,
        crossOrigin: ctx.mext.config.crossOrigin,
        error: ctx.mext.error,
        modules: ctx.mext.modules
      }),
      renderReactToString
    );
  } catch (error) {
    ctx.mext.error = true;
    throw error;
  } finally {
    ctx.renderTime = Number(new Date()) - _startTime;
  }
};
