// mext/lib/modules/MultiplePage/main/utils/fetch-root.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
/**
 * 这里是服务于单页面应用初始化页面时候的函数实现
 */

const mext_utils_1 = require('mext-utils');

const client_1 = require('mext-plugin/client');

const setMext_1 = require('../../../../setMext');

exports.default = async (getInitialProps, INITIAL_STATE, routes) => {
  if (getInitialProps) {
    const searchInfo = mext_utils_1.search();
    const query = mext_utils_1.queryObj(searchInfo);

    try {
      /**
       * 初始化非路由组件的数据，即挂载在入口组件的getInitialProps方法的实现
       */
      const initialPropsParams = {
        location: window.location,
        query,
        setMext: setMext_1.default,
        routes,
        route: null,
        match: null
      };
      await client_1.default.hooks.modifyInitialPropsCtx({
        params: initialPropsParams
      });
      const rootProps = await getInitialProps(initialPropsParams);
      INITIAL_STATE.mext = { ...INITIAL_STATE.mext, ...rootProps };
    } catch (err) {
      client_1.default.hooks.catchError({
        type: 'fetch',
        error: err
      });
      INITIAL_STATE.MextException = await mext_utils_1.Exception.handleError.call(null, err);
    }
  }
};
