// mext-server/lib/utils/resource.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
/**
 * 生产环境初始化所有资源
 */

const fs = require('fs-extra');

const path_1 = require('path');

const immutable_1 = require('immutable');

const Loadable = require('react-loadable');

let isReadPublicMap = false;

async function default_1() {
  if (!this.apiServer && !this.dev) {
    if (!isReadPublicMap) {
      isReadPublicMap = true;
      const config = this.config.toObject();
      const { server_dist, entry, mode } = config; // 读mapFile、manifestFile, documentFile

      const mapJson = path_1.join(this.dir, server_dist, '.mextConfig/map.json');
      const manifestDir = path_1.join(this.dir, server_dist, '.mextConfig/manifest.js');
      const documentDir = path_1.join(this.dir, server_dist, 'document.js');
      const ico = path_1.join(this.dir, 'favicon.ico'); // 读取资源映射名称

      if (fs.existsSync(mapJson)) {
        // 保证数据源数据不变性
        this.map = immutable_1.Seq(JSON.parse(fs.readFileSync(mapJson, 'utf-8')));
      } // 读取manifest文件

      if (fs.existsSync(manifestDir)) {
        this.manifestFile = fs.readFileSync(manifestDir, 'utf-8');
      } // 读取favicon.ico

      if (fs.existsSync(ico)) {
        this.favicon = fs.readFileSync(ico);
      } // require  document组件

      if (fs.existsSync(documentDir)) {
        const DocumentComponent = require(documentDir);

        this.RootDocumentComponent = DocumentComponent.default || DocumentComponent;
      } // 获取入口文件地址

      this.RootPageEntry = path_1.join(this.dir, server_dist, entry.replace(/\.tsx?$/, '.js'));

      if (mode === 'server') {
        // 引用入口直接执行AppRegistry
        require(this.RootPageEntry);

        await Loadable.preloadAll();

        if (this.RootComponent === null) {
          // 说明没有执行start方法
          throw new Error('请确认入口已经使用start方法启动了');
        } // 解析路由规则，获取当前项目的所有路由列表
        // 保证数据源数据不变性

        this.routes = immutable_1.fromJS(global.__MEXT__INIT__ROUTES__ || []);
      }
    }
  }
}

exports.default = default_1;
